docs/zkapps/o1js/bitwise-gadgets.mdx

```diff
@@ -0,0 +1,320 @@
+---
+title: Bitwise Gadgets
+hide_title: true
+description: Bitwise gadgets in o1js.
+keywords:
+  - gadgets
+  - bitwise
+  - o1js
+---
+
+:::info
+
+Please note that zkApp programmability is not yet available on Mina Mainnet, but zkApps can now be deployed to Berkeley Testnet.
+
+:::
+
+# Bitwise Gadgets in o1js
+
+Bitwise operations work on 2-bit patterns of equal lengths by positionally matching their individual bits.
+
+- [and](#and)
+- [compactMultiRangeCheck](#compactmultirangecheck)
+- [leftShift](#leftshift)
+- [multiRangeCheck](#multirangecheck)
+- [not](#not)
+- [rangeCheck64](#rangecheck64)
+- [rightShift](#rightshift)
+- [rotate](#rotate)
+- [xor](#xor)
+
+## and
+
+```ts
+and(a: Field, b: Field, length: number) {
+    return and(a, b, length);
+```
+
+The bitwise and gadget on Field elements is equivalent to the [bitwise AND (&)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_AND) operator in JavaScript. 
+
+- The and gadget works by comparing two bits and returns `1` only if both bits are `1`, otherwise it returns `0`.
+- Constrains both input values to fit into `paddedLength` bits. 
+- Both Field elements must fit into `2^paddedLength - 1` or an error is thrown.
+- Guarantees that the output is has at most `paddedLength` bits.
+- Can be checked by a double generic gate that verifies the following relationship between the values. In the process, it also invokes the [xor()](#xor) gadget that creates additional constraints depending on `length`.
+
+  The generic gate verifies `a + b = sum` and the conjunction equation `2 * and = sum - xor`.
+
+  where:
+  * `a + b = sum`
+  * `a ^ b = xor`
+  * `a & b = and`
+
+For details about the implementation, see [AND](https://o1-labs.github.io/proof-systems/specs/kimchi.html?highlight=gates#and) in the Mina book.
+
+The `length` parameter:
+
+- Defines how many bits to compare. 
+- Rounds `length` to the nearest multiple of 16, `paddedLength = ceil(length / 16) * 16`.
+- Specifying a larger `length` parameter adds additional constraints.
+
+With `length = 2` (`paddedLength = 16`), the `and()` operation fails for any input that is larger than `2**16`.
+
+Example:
+
+```ts
+let a = Field(3);    // ... 000011
+let b = Field(5);    // ... 000101
+let c = Gadgets.and(a, b, 2);    // ... 000001
+c.assertEquals(1);
+```
+
+## compactMultiRangeCheck
+
+```ts
```

Comment:
we can provide the syntax for each gadget... not sure I have all of them right, but sometimes all a dev needs is the syntax 
(do we want to document the parameters?)